### PR TITLE
Whitelist URL class

### LINKED ISSUE #60660: Painless convenience functions for parsing URLs
Perhaps either whitelist [URL](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html) or wrapping that function.

Should be able to get scheme, host, port, path, fragment.

Meta: https://github.com/elastic/elasticsearch/issues/60659

### PR DESCRIPTION
**Pull request by interpreter95**
Exposes parts of the [java.net.URL](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html) class in painless.

API:
```
class java.net.URL {
    (String)
    String getHost()
    String getPath()
    int getPort()
    String getProtocol()
    String getRef()
}
```

Closes: #60660

### CODE REVIEW

**Comment by interpreter95** [comment]
I just want to make the reviewer aware that when I ran `./gradlew check` locally I seem to fail `:libs:elasticsearch-ssl-config:test` with error message `elasticsearch Cannot find trusted issuer with name [thawte]`. I assume that this failure is because of a configuration error and not part of my commit. If not, please let me know and I will debug further!

**Comment by stu-elastic** [comment]
@elasticmachine ok to test

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/UrlTests.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        e = expectScriptThrows(MalformedURLException.class, () -> exec("new URL(null)"));
+        assertEquals("Cannot invoke \"String.length()\" because \"spec\" is null", e.getMessage());
+
+        e = expectScriptThrows(MalformedURLException.class, () -> exec("new URL('abc://www.test.com')"));
+        assertEquals("unknown protocol: abc", e.getMessage());
+    }
+
+    public void testGetHost() {
+        assertEquals("www.test.com", exec("URL url = new URL('https://www.test.com:8080/path?query#fragment'); url.getHost()"));
```
Comment (stu-elastic) [inline modules/lang-painless/src/test/java/org/elasticsearch/painless/UrlTests.java:29]: Please use `example.com`

**Comment by stu-elastic** [comment]
We're going to have to override `hashCode()` and `equals()` because they are blocking operations.

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/UrlTests.java (hunk 1 above)
Comment (interpreter95) [reply modules/lang-painless/src/test/java/org/elasticsearch/painless/UrlTests.java]: Will do!

**Comment by interpreter95** [comment]
Hi,

I see in the Jenkins logs that my pull request is failing tests because of:

```
15:21:49 runbld>>> >>>>>>>>>>>> SCRIPT EXECUTION BEGIN >>>>>>>>>>>>
15:21:49 runbld>>> Adding /var/lib/jenkins/.java/openjdk-15.0.2+7-linux/bin to the path.
15:21:49 + /dev/shm/elastic+elasticsearch+master+pull-request+bwc/.ci/scripts/run-gradle.sh -Dignore.tests.seed bwcTestSnapshots
15:21:49 /tmp/jenkins1340635122548236278.sh: line 2: /dev/shm/elastic+elasticsearch+master+pull-request+bwc/.ci/scripts/run-gradle.sh: No such file or directory
15:21:49 runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
```

Seems like a file or folder has been moved... But from what I know, I have only modified/added files in my pull request.

Anyone have any hints on what I should do to fix this?

**Comment by rjernst** [comment]
> Anyone have any hints on what I should do to fix this?

Our CI infrastructure changed. You'll need to sync with the latest from the `master` branch.

**Comment by stu-elastic** [comment]
@interpreter95 We're going to have to figure out a way to automatically specify a [URLStreamHandler](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#URL(java.net.URL,%20java.lang.String,%20java.net.URLStreamHandler)) that performs no-ops. 

Otherwise the [system](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#URL(java.lang.String,%20java.lang.String,%20int,%20java.lang.String)) `URLStreamHandler` is used for the given protocol.  We can't have `equals` and `hashcode` be blocking network operations and a dummy URLStreamHandler is the best way to avoid any surprises.

After we've solved that issue, we can modify this PR to get it merged.

**Comment by interpreter95** [comment]
Thanks for the update @stu-elastic 😃 

I'm just cleaning up the code a bit, and also fixing a bug in my tests that the latest test of the PR found!

Feel free to update this PR as you like, I just wanted to leave it in a state where it's hopefully easy for you to modify it.

I will submit an update to the PR today just to check that all the tests pass. Then I leave it up to you to decide what to do with the PR.

**Comment by stu-elastic** [comment]
Hi @interpreter95, as mentioned in https://github.com/elastic/elasticsearch/issues/60660#issuecomment-875808518, we're going to go another way for this task.  This is the correct way to whitelist and the correct way to test whitelisted methods.  I hope you'll pick another painless issue to work on.  Feel free to ping me on whichever issue you pick up next and I'll endeavor to be more responsive.

Thanks.

**Pull request closed without merging**